### PR TITLE
Enhanced Subnet Discovery

### PR DESCRIPTION
**Pull request by dshehbaj**
add testing doc here

add integration test screenshot here

add edge cases here

**What type of PR is this?**

**Which issue does this PR fix?**:

**What does this PR do / Why do we need it?**:

**Testing done on this change**:

**Will this PR introduce any new dependencies?**:

**Will this break upgrades or downgrades? Has updating a running cluster been tested?**:

**Does this change require updates to the CNI daemonset config files to work?**:

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/awsutils/awsutils.go [hunk 1]
```diff
@@ -966,28 +969,65 @@ func (cache *EC2InstanceMetadataCache) createENI(useCustomCfg bool, sg []*string
...
+			if checkErr != nil {
+				// If we can't determine exclusion status, log warning and proceed
+				log.Warnf("Failed to check if primary subnet is excluded: %v. Proceeding with ENI creation attempt.", checkErr)
+			} else if excluded {
+				// Primary subnet is explicitly excluded
+				err = errors.New("primary subnet is tagged with kubernetes.io/role/cni=0 and subnet discovery is disabled - no valid subnets available for ENI creation")
+				log.Error(err.Error())
+				return "", err
+			}
+
```
Comment (dshehbaj) [inline pkg/awsutils/awsutils.go:1179]: Deals with misconfiguration scenario where customer tags subnet for exclusion but does not have subnet discovery enabled.

File: pkg/awsutils/awsutils.go [hunk 2]
```diff
@@ -966,28 +969,65 @@ func (cache *EC2InstanceMetadataCache) createENI(useCustomCfg bool, sg []*string
...
+				// Even in fallback, check if primary subnet is excluded
+				excluded, checkErr := cache.isPrimarySubnetExcluded()
+				if checkErr != nil {
+					log.Warnf("Failed to check if primary subnet is excluded: %v. Proceeding with ENI creation attempt.", checkErr)
+				} else if excluded {
+					// Primary subnet is explicitly excluded
+					err = errors.New("primary subnet is tagged with kubernetes.io/role/cni=0 - no valid subnets available for ENI creation")
+					log.Error(err.Error())
+					return "", err
+				}
```
Comment (dshehbaj) [inline pkg/awsutils/awsutils.go:1122]: Dealing with the case where describe subnet call fails, but customer has also opted to exclude the primary subnet.

File: pkg/ipamd/datastore/data_store.go [hunk 3]
```diff
@@ -797,6 +843,11 @@ func (ds *DataStore) GetIPStats(addressFamily string) *DataStoreStats {
 		TotalPrefixes: ds.allocatedPrefix,
 	}
 	for _, eni := range ds.eniPool {
+		// Skip excluded ENIs when calculating available IPs for pod allocation
+		if eni.IsExcludedForPodIPs {
+			continue
+		}
+
```
Comment (dshehbaj) [inline pkg/ipamd/datastore/data_store.go:850]: We skip the excluded ENI (which would the Primary ENI in this case practically speaking) from IP stat calculation. 

Doing it here would automatically take care of warm target calculation in other places such as `isDatastorePoolTooLow`, `shouldRemoveExtraENIs`.

File: pkg/awsutils/awsutils.go [hunk 4]
```diff
@@ -506,6 +510,205 @@ func (cache *EC2InstanceMetadataCache) initWithEC2Metadata(ctx context.Context)
...
+		awsAPIErrInc("DiscoverCustomSecurityGroups", err)
+		log.Warnf("Failed to discover custom security groups: %v", err)
+		log.Info("Falling back to using primary security groups for ENIs in secondary subnets")
+
+		// Clear custom security groups cache to trigger fallback behavior
+		cache.customSecurityGroups.Set([]string{})
+
+		// Apply primary security groups to ENIs in secondary subnets as fallback
+		return cache.applyFallbackSecurityGroups(store)
+	}
```
Comment (dshehbaj) [inline pkg/awsutils/awsutils.go]: Fallback to primary security group and show logs to customer if we fail to describe any secondary security groups.

We do the same fallback for primary subnets.
